external-crates/move/crates/move-package/tests/test_sources/resolve_pkg_name_conflict_lockfile/Move.toml

```diff
@@ -0,0 +1,22 @@
+# This test highlights the problem of error messages using resolved pkg
+# names instead of original when dependnecy lock files are present. The
```

Comment:
It solved it largely, but we don't store the original name in the lockfile, so when loding the graph from the lockfile the resolved name is used (this is just for error messages though). It can only be solved by adding original pkg name to lockfiles which is actually a discussion point (3) in the OP #14178

Reply:
> There's also a confusion here in that if there is a conflict between an on-chain dependency and a source dependency, then the only identity they share is the publish-at address, which is the resolved name

So just to clarify this, we have:
- original name (user-friendly name in the manifest)
- resolved name (in Sui's case this will be the original package id)
- resolved version (in Sui's case this will be the published-at id)

So when the algorithm is resolving the graph, it will load all packages and store them in the table based on their `resolved name` as the key. And when it encounters another package whose `resolved name` already exists in the table, it will then compare both packages based on their `resolved version` (which is implicitly done here https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L717) and their dependencies (done immediately below https://github.com/MystenLabs/sui/blob/main/external-crates/move/crates/move-package/src/resolution/dependency_graph.rs#L751)

Reply:
Based on https://github.com/MystenLabs/sui/pull/15867/files#r1470220325 we're interested in keeping the `resolved name` in the lock file; that has my vote for readability and convenience.

Reply:
Maybe a dumb question: what happens if a user duplicately specify an on-chain dependency?

Reply:
Same thing that happens when you specify a duplicate source dependency. Both on-chain and source dependencies get resolved to `(original_id, published_at)` so after that it doesn't matter to the resolver where the dependency came from.
There are some rare edge cases where transitive deps pulled in from different locations do and don't define their published-at which is briefly dicussed [here](https://gist.github.com/kklas/d9edad7875548fae57b0c86bc8871fea) in the "Backwards compatibility" section  (rarely should happen and you can use overrides to go around it).